gbpservice/neutron/nsf/configurator/agents/firewall.py

```diff
@@ -0,0 +1,323 @@
+        self._sc.rpc_event(ev)
+
+    def delete_firewall(self, context, firewall, host):
+        LOG.debug("FwaasRpcReceiver received Delete Firewall request.")
+        arg_dict = {'context': context,
+                    'firewall': firewall,
+                    'host': host}
+        ev = self._sc.event(id='DELETE_FIREWALL', data=arg_dict)
+        self._sc.rpc_event(ev)
+
+
+class FWaasEventHandler(object):
+    """
+    Handler class for demultiplexing firewall configuration
+    requests from Fwaas Plugin and sending to appropriate driver.
+    """
+
+    def __init__(self, sc, drivers, rpcmgr, nqueue):
+        self._sc = sc
+        self.drivers = drivers
+        self.host = cfg.CONF.host
+        self._rpcmgr = rpcmgr
+        self.plugin_rpc = FwaasRpcSender(
+            sc, self.host, nqueue)
+
+    def _get_driver(self):
+        driver_id = const.SERVICE_TYPE
+        return self.drivers[driver_id]
+
+    def _is_firewall_rule_exists(self, fw):
+        if not fw['firewall_rule_list']:
+            return False
+        else:
+            return True
+
+    def handle_event(self, ev):
+        try:
+            msg = ("Worker process with ID: %s starting to "
+                   "handle task: %s of type firewall. "
+                   % (os.getpid(), ev.id))
+            LOG.debug(msg)
+
+            driver = self._get_driver()
+            self.method = getattr(driver, "%s" % (ev.id.lower()))
+            self.invoke_driver_for_plugin_api(ev)
+        except Exception as err:
+            LOG.error("Failed to perform the operation: %s. %s"
+                      % (ev.id, str(err).capitalize()))
+
+    def invoke_driver_for_plugin_api(self, ev):
+        context = ev.data.get('context')
+        firewall = ev.data.get('firewall')
+        host = ev.data.get('host')
+
+        if ev.id == 'CREATE_FIREWALL':
+            if not self._is_firewall_rule_exists(firewall):
+                msg = ("Firewall status set to ACTIVE")
+                LOG.debug(msg)
+                return self.plugin_rpc.set_firewall_status(
+                    context, firewall['id'], const.STATUS_ACTIVE)
+            # Added to handle in service vm agents. VM agent will add
+            # default DROP rule.
+            # if not self._is_firewall_rule_exists(firewall):
+            #     self.plugin_rpc.set_firewall_status(
+            #         context, firewall['id'], const.STATUS_ACTIVE)
+            try:
+                status = self.method(context, firewall, host)
+            except Exception as err:
+                self.plugin_rpc.set_firewall_status(
+                    context, firewall['id'], const.STATUS_ERROR)
+                msg = ("Failed to configure Firewall and status is "
+                       "changed to ERROR. %s." % str(err).capitalize())
+                LOG.error(msg)
+            else:
+                self.plugin_rpc.set_firewall_status(
+                    context, firewall['id'], status)
```

Comment:
When notification is sent, we are sending "success"/<Failure string> back to orchestrator. Will this status code match with this or we need to convert?